src/Symfony/Component/Messenger/Bridge/Kafka/Transport/KafkaReceiver.php

```diff
@@ -0,0 +1,151 @@
+use Symfony\Component\Messenger\Transport\Receiver\ReceiverInterface;
+use Symfony\Component\Messenger\Transport\Serialization\SerializerInterface;
+
+/**
+ * @author Konstantin Scheumann <konstantin@konstantin.codes>
+ */
+class KafkaReceiver implements ReceiverInterface
+{
+    /** @var LoggerInterface */
+    private $logger;
+
+    /** @var SerializerInterface */
+    private $serializer;
+
+    /** @var RdKafkaFactory */
+    private $rdKafkaFactory;
+
+    /** @var KafkaReceiverProperties */
+    private $properties;
+
+    /** @var KafkaConsumer */
+    private $consumer;
+
+    /** @var bool */
+    private $subscribed;
+
+    public function __construct(
+        LoggerInterface $logger,
+        SerializerInterface $serializer,
+        RdKafkaFactory $rdKafkaFactory,
+        KafkaReceiverProperties $properties
+    ) {
+        $this->logger = $logger;
+        $this->serializer = $serializer;
+        $this->rdKafkaFactory = $rdKafkaFactory;
+        $this->properties = $properties;
+
+        $this->subscribed = false;
+    }
+
+    public function get(): iterable
+    {
+        $message = $this->getSubscribedConsumer()->consume($this->properties->getReceiveTimeoutMs());
+
+        switch ($message->err) {
+            case \RD_KAFKA_RESP_ERR_NO_ERROR:
+                $this->logger->info(sprintf(
+                    'Kafka: Message %s %s %s received ',
+                    $message->topic_name,
+                    $message->partition,
+                    $message->offset
+                ));
+
+                $envelope = $this->serializer->decode([
+                    'body' => $message->payload,
+                    'headers' => $message->headers,
+                    'key' => $message->key,
+                    'partition' => $message->partition,
+                    'offset' => $message->offset,
+                    'timestamp' => $message->timestamp,
+                ]);
+
+                return [$envelope->with(new KafkaMessageStamp($message))];
+            case \RD_KAFKA_RESP_ERR__PARTITION_EOF:
+                $this->logger->info('Kafka: Partition EOF reached. Waiting for next message ...');
+                break;
+            case \RD_KAFKA_RESP_ERR__TIMED_OUT:
+                $this->logger->debug('Kafka: Consumer timeout.');
+                break;
+            case \RD_KAFKA_RESP_ERR__TRANSPORT:
+                $this->logger->debug('Kafka: Broker transport failure.');
+                break;
+            default:
+                throw new TransportException($message->errstr(), $message->err);
+        }
+
+        return [];
+    }
+
+    public function ack(Envelope $envelope): void
+    {
+        $consumer = $this->getConsumer();
+
+        /** @var KafkaMessageStamp $transportStamp */
+        $transportStamp = $envelope->last(KafkaMessageStamp::class);
+        $message = $transportStamp->getMessage();
+
+        if ($this->properties->isCommitAsync()) {
+            $consumer->commitAsync($message);
```

Comment:
I'm not comfortable with providing async `ack`. it means you don't know if the message has been acked (ie. transport failure), but Symfony will not be aware of that.

Reply:
That's true. It would only be acceptable in a "fire and forget" scenario.
We could work with a callback here, that creates a debug-level log after offset commit?

Reply:
Using `commit` (synchronous) usually drops the throughput of the implementation (afaik). From my experience people that use phprdkafka usually switch to async commits strictly because the difference in speed it has.

@nick-zh can you support / deny me on this?

Reply:
Yes there is an overhead and depending on how fast you need to read, this might throttle your throughput. I would opt for a setting like: commit after every x messages
This way you would have the "safety" of the sync commit but throughput doesn't need to suffer.
Low throughput users could still opt for sync commit after every message, which probably is desirable.
We'll be doing something similar with the new consumers we'll introduce.
Handling async commits is also a valid option, but yeah you would need some more flexibility to handle failures, etc.
Sometimes this can be a bother and adds complexity of course.
I hope this helps, sry i am not really familiar with this implementation :sweat_smile:

Reply:
> Using `commit` (synchronous) usually drops the throughput of the implementation (afaik). From my experience people that use phprdkafka usually switch to async commits strictly because the difference in speed it has.
> 
> @nick-zh can you support / deny me on this?

From my little experience, throughput drops dramatically with sync commits. But, performance loss may be reduced by using multiple consumers and partitions. Unfortunately, I can’t provide you benchmark data for now… But maybe soon.